spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/HealthCheckServiceInstanceListSupplier.java

```diff
@@ -0,0 +1,136 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.loadbalancer.core;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+import org.springframework.cloud.client.ServiceInstance;
+import org.springframework.cloud.client.loadbalancer.reactive.LoadBalancerProperties;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.reactive.function.client.WebClient;
+import org.springframework.web.util.UriComponentsBuilder;
+
+/**
+ * A {@link ServiceInstanceListSupplier} implementation that verifies whether the
+ * instances are alive and only returns the healthy one, unless there are none. Uses
+ * {@link WebClient} to ping the <code>health</code> endpoint of the instances.
+ *
+ * @author Olga Maciaszek-Sharma
+ * @since 2.2.0
+ */
+public class HealthCheckServiceInstanceListSupplier
+		implements ServiceInstanceListSupplier {
+
+	private static final Log LOG = LogFactory
+			.getLog(HealthCheckServiceInstanceListSupplier.class);
+
+	private final ServiceInstanceListSupplier delegate;
+
+	private final LoadBalancerProperties.HealthCheck healthCheck;
+
+	private final WebClient webClient;
+
+	private final String defaultHealthCheckPath;
+
+	private List<ServiceInstance> instances = Collections
+			.synchronizedList(new ArrayList<>());
+
+	private List<ServiceInstance> healthyInstances = Collections
+			.synchronizedList(new ArrayList<>());
+
+	public HealthCheckServiceInstanceListSupplier(ServiceInstanceListSupplier delegate,
+			LoadBalancerProperties.HealthCheck healthCheck, WebClient webClient) {
+		this.delegate = delegate;
+		this.healthCheck = healthCheck;
+		defaultHealthCheckPath = healthCheck.getPath().getOrDefault("default",
+				"/actuator/health");
+		this.webClient = webClient;
+		initInstances();
+
+	}
+
+	private void initInstances() {
+		delegate.get().subscribe(delegateInstances -> {
+			instances.clear();
+			instances.addAll(delegateInstances);
+		});
+
+		Flux<List<ServiceInstance>> healthCheckFlux = healthCheckFlux();
+
+		healthCheckFlux.subscribe(verifiedInstances -> {
+			healthyInstances.clear();
+			healthyInstances.addAll(verifiedInstances);
+		});
+	}
+
+	protected Flux<List<ServiceInstance>> healthCheckFlux() {
+		return Flux.create(emitter -> Schedulers
```

Comment:
@OlgaMaciaszek @spencergibb  @ryanjbaxter 
IMO `Flux.create` does looks like WA  
This could be rewritten without usage of it and `Scheduler#schedulePeriodically` and more Thread safely 
```java
protected Flux<List<ServiceInstance>> healthCheckFlux() {
	return Flux.defer(() -> {
		List<ServiceInstance> result = new CopyOnWriteArrayList<>();
		List<Mono<ServiceInstance>> checks = new ArrayList<>();
		for (ServiceInstance instance : instances) {
			Mono<ServiceInstance> alive = isAlive(instance)
					.timeout(healthCheck.getInterval(), Mono.empty())//ensures that batch will be produced no longer than healthCheck.getInterval()
					.filter(it -> it)
					.map(check -> instance);
			checks.add(alive);
		}
		return Flux.merge(checks)//executes in parallel
				.map(alive -> {//will emit in on Each healthy instance with saving result of the current batch check 
					result.add(alive);
					return result;
				})
				.defaultIfEmpty(result);
	})
	.repeatWhen(restart -> restart.delayElements(healthCheck.getInterval()))//
	.delaySubscription(Duration.ofMillis(healthCheck.getInitialDelay()));
}

private void initInstances() {
	healthCheckDisposable = delegate.get()
			.doOnNext(delegateInstances -> {
				instances = Collections.unmodifiableList(new ArrayList<>(delegateInstances));
			})
			.thenMany(healthCheckFlux())//chains operations as they depends on each other
			.subscribeOn(Schedulers.parallel())//subscribes async 
			.subscribe(verifiedInstances -> healthyInstances.set(verifiedInstances));//atomic operation 
}

@Override
public void destroy() {
	this.healthCheckDisposable.dispose(); // disposes task so bean could be destroyed without leaks 
}
```